File: src-tauri/src/app/commands.rs
//! Frontend-facing commands: thin `#[tauri::command]` wrappers over `ops`.
//! Function names are the command names the frontend invokes (design §commands).

use std::sync::Arc;

use tauri::State;

use super::autostart::PluginAutostart;
use super::events::TauriSink;
use super::hotkey::{HotkeyState, PluginRegistrar};
use super::ops;
use super::platform;
use super::state::AppState;
use crate::model::{
    AppError, ItemDto, ItemId, PlatformInfo, Settings, TransferMode, TransferOutcome,
    TransferPreview,
};

pub type SharedState = Arc<AppState>;

#[tauri::command]
pub fn list_items(state: State<'_, SharedState>) -> Vec<ItemDto> {
    ops::list_items(&state)
}

#[tauri::command]
pub fn transfer_item(
    state: State<'_, SharedState>,
    id: ItemId,
    mode: TransferMode,
) -> Result<TransferOutcome, AppError> {
    ops::transfer_item(&state, id, mode)
}

#[tauri::command]
pub fn preview_transfer(
    state: State<'_, SharedState>,
    id: ItemId,
) -> Result<TransferPreview, AppError> {
    ops::preview_transfer(&state, id)
}

#[tauri::command]
pub fn remove_item(
    app: tauri::AppHandle,
    state: State<'_, SharedState>,
    id: ItemId,
) -> Result<(), AppError> {
    ops::remove_item(&state, &TauriSink(app), id)
}

#[tauri::command]
pub fn clear_items(app: tauri::AppHandle, state: State<'_, SharedState>) {
    ops::clear_items(&state, &TauriSink(app));
}

#[tauri::command]
pub fn get_settings(state: State<'_, SharedState>) -> Settings {
    ops::get_settings(&state)
}

/// Validate, register a changed hotkey, persist, apply the diff and announce (9.3, 9.4).
#[tauri::command]
pub fn set_settings(
    app: tauri::AppHandle,
    state: State<'_, SharedState>,
    hotkeys: State<'_, HotkeyState>,
    settings: Settings,
) -> Result<Settings, AppError> {
    ops::apply_settings(
        &state,
        &hotkeys,
        &PluginRegistrar(app.clone()),
        &PluginAutostart(app.clone()),
        &TauriSink(app),
        settings,
    )
}

#[tauri::command]
pub fn get_platform_info(state: State<'_, SharedState>) -> PlatformInfo {
    platform::platform_info(state.clipboard.as_ref())
}

#[tauri::command]
pub fn hide_window(app: tauri::AppHandle) {
    super::window::hide(&app);
}

#[tauri::command]
pub fn open_settings(app: tauri::AppHandle) -> Result<(), AppError> {
    super::window::open_settings(&app)
        .map_err(|_| AppError::from(crate::model::ErrorKind::SettingsIo))
}

File: src/lib/ipc/commands.ts
/**
 * Typed wrappers over Tauri `invoke`. This module (and `events.ts`) are the only places
 * that import `@tauri-apps/api`; components go through `commands` so the IPC surface stays
 * in one file and can be faked in tests.
 *
 * Rejections are `AppError` objects (`{ kind }`) when the backend refused the call; anything
 * else is passed through unchanged.
 */

import { invoke } from '@tauri-apps/api/core';
import type {
  ItemDto,
  ItemId,
  PlatformInfo,
  Settings,
  TransferMode,
  TransferOutcome,
  TransferPreview,
} from './types';

export const commands = {
  listItems: (): Promise<ItemDto[]> => invoke('list_items'),
  transferItem: (id: ItemId, mode: TransferMode): Promise<TransferOutcome> =>
    invoke('transfer_item', { id, mode }),
  previewTransfer: (id: ItemId): Promise<TransferPreview> => invoke('preview_transfer', { id }),
  removeItem: (id: ItemId): Promise<void> => invoke('remove_item', { id }),
  clearItems: (): Promise<void> => invoke('clear_items'),
  getSettings: (): Promise<Settings> => invoke('get_settings'),
  setSettings: (settings: Settings): Promise<Settings> => invoke('set_settings', { settings }),
  getPlatformInfo: (): Promise<PlatformInfo> => invoke('get_platform_info'),
  hideWindow: (): Promise<void> => invoke('hide_window'),
  openSettings: (): Promise<void> => invoke('open_settings'),
  closeSettings: (): Promise<void> => invoke('close_settings'),
  closeAbout: (): Promise<void> => invoke('close_about'),
  suspendHotkey: (): Promise<void> => invoke('suspend_hotkey'),
  resumeHotkey: (): Promise<void> => invoke('resume_hotkey'),
} as const;

export type Commands = typeof commands;
